### PR TITLE
[flang] Validate DIM argument for reduction intrinsics that return a …

### PR DESCRIPTION
**Pull request by psteinfeld**
…scalar

Reduction intrinsics that have an ARRAY argument of rank 1 can only accept a DIM argument whose value is 1.  We were not detecting cases where DIM was not equal to 1.  Here's a small program that illustrates the problem:
  program minval_test
    integer :: array(3) = [1,2,3]
    integer :: dim
    dim = 0
    print *, minval(array, dim)
  end program minval_test

Our implementation of reduction intrinsics has two sets of interfaces -- one set for intrinsics that return an array and another set for intrinsics that return a scalar.  Some of the intrinsics that return a scalar stem from calls without the DIM argument.  Others stem from calls with the DIM argument where the ARRAY argument has rank 1.  But for scalar-returning calls, we were not passing along the information of whether the call contained a DIM argument.

I fixed this by adding a boolean argument to the interface to the runtime that indicates whether the original source contained a DIM argument.  This small change propatated through a lot of the runtime resulting in several files changing.

This is an initial implementation that only implements this change for the MINVAL intrinsic.  If I receive initial comments validating this approach, I'll expand it to the other reduction intrinsics.

### CODE REVIEW

File: flang/unittests/Optimizer/Builder/Runtime/ReductionTest.cpp [hunk 1]
```diff
@@ -89,27 +90,63 @@ TEST_F(RuntimeCallTest, genMaxValTest) {
...
+  mlir::Value min = fir::runtime::genMinval(builder, loc, undef, false, undef,
+      mask);
+  checkCallOp(min.getDefiningOp(), fctName, 4);
 }
 
 TEST_F(RuntimeCallTest, genMinValTest) {
+  FILE *file;
+  file = fopen("/home/psteinfeld/log", "a");
+  fprintf(file, "In genMinValTest\n");
+  fclose(file);
```
Comment (jeanPerier) [inline flang/unittests/Optimizer/Builder/Runtime/ReductionTest.cpp:112]: It looks like some debug code was left here (and the test is failing)

**Review by jeanPerier — COMMENTED** [review]
Given `DIM` cannot be dynamically optional, the approach of passing a boolean + integer value looks reasonable to me. Please get @klausler opinion here since this is mostly a runtime API change that could be pushed to LLVM directly.

File: flang/unittests/Optimizer/Builder/Runtime/ReductionTest.cpp (hunk 1 above)
Comment (kiranchandramohan) [reply flang/unittests/Optimizer/Builder/Runtime/ReductionTest.cpp:112]: @psteinfeld I was able to reproduce the failure locally on my Mac. I guess Mac's filesystem layout is different and also you are specifically referring to your home directory in linux(?).
After removing the debug log messages and the commented code, the test passes for me.

Comment (PeteSteinfeld) [reply flang/unittests/Optimizer/Builder/Runtime/ReductionTest.cpp:112]: Thanks, @kiranchandramohan.  You're the best.

**Comment by PeteSteinfeld** [comment]
> Given `DIM` cannot be dynamically optional, the approach of passing a boolean + integer value looks reasonable to me. Please get @klausler opinion here since this is mostly a runtime API change that could be pushed to LLVM directly.

Thanks, @jeanPerier, will do.

**Comment by klausler** [comment]
Better: add an API to the runtime that just tests its argument for equality to 1 and crashes when it's not.  In optimizing compilations, don't make the call at all.  Fortran allows a compiler to omit calculations whose results are not needed to compute the result of an expression.

**Comment by psteinfeld** [comment]
On the advice of several people, I'm abandoning this pull request.

**Pull request closed without merging**